Sources/SwiftConcurrentSequence/Array+ConcurrentMap.swift

```diff
@@ -0,0 +1,70 @@
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS"BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Dispatch
+
+extension Array {
+    /// Returns an array containing the results of mapping the given closure
+    /// over the sequence's elements. The given closure is executed concurrently
+    /// on multiple queues to reduce the wall-time consumed by the transform.
+    ///
+    /// - Parameter transform: A mapping closure. `transform` accepts an
+    ///   element of this sequence as its parameter and returns a transformed
+    ///   value of the same or of a different type.
+    /// - Returns: An array containing the transformed elements of this
+    ///   sequence.
+    public func concurrentMap<T>(_ transform: (Element) -> T) -> [T] {
+        // Create a buffer where we can store the transformed output.
+        var transformed = [T?](repeating: nil, count: count)
+        // Access the underlying array memory to concurrently write to indices.
+        return transformed.withUnsafeMutableBufferPointer { bufferPointer in
+            DispatchQueue.concurrentPerform(iterations: count) { index in
+                // It is safe to concurrently write to unique indexes of the buffer pointer.
+                bufferPointer[index] = transform(self[index])
+            }
+            return bufferPointer.compactMap { $0 }
+        }
+    }
+
+    /// Returns an array containing the results of mapping the given closure
+    /// over the sequence's elements. The given closure is executed concurrently
+    /// on multiple queues to reduce the wall-time consumed by the transform.
+    ///
+    /// - Parameter transform: A mapping closure. `transform` accepts an
+    ///   element of this sequence as its parameter and returns a transformed
+    ///   value of the same or of a different type.
+    /// - Returns: An array containing the transformed elements of this
+    ///   sequence.
+    public func concurrentMap<T>(_ transform: @escaping (Element) throws -> T) async rethrows -> [T] {
+        try await withThrowingTaskGroup(
+            of: (index: Int, element: T).self,
+            returning: [T].self)
+        { group in
+            for index in 0..<count {
+                group.addTask {
+                    (index, try transform(self[index]))
+                }
+            }
+
+            var transformed = [T?](repeating: nil, count: count)
+            for try await transform in group {
+                transformed[transform.index] = transform.element
```

Comment:
I'm starting by trying to create a version of `map` and `reduce` that behave in a similar manner to their non-concurrent counterparts. But I do like the the idea of creating performance-optimized, concurrent, mutable method to `remap` (needs a different name than `map` I think?) existing `RandomAccessCollection`s.

I can't create that api on top of `RandomAccessCollection` though without `await` since there is no `withUnsafeMutableBufferPointer` on that protocol.

Reply:
Ah, fascinating. Well using a dispatch group or some such to signal completion would certainly let me better compare perf.

Reply:
Tried to get this working but passing data between structured concurrent and unstructured concurrent code is quite a challenge (by design!). Going to punt on this for this PR but it's something I'll keep iterating on.

That said, some naive performance analysis did indicate that the `concurrentPerform` code was ~5x faster than the `async` code. So I'm not sure what the ROI is from continuing down the path of trying to make the synchronous code utilize the async paths.